src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala

```diff
@@ -266,28 +279,37 @@ object RuleUtils {
       index: IndexLogEntry,
       plan: LogicalPlan,
       useBucketSpec: Boolean): LogicalPlan = {
+
+    def makeHadoopFsRelation(index: IndexLogEntry, relation: LogicalPlan) = {
+      val location = index.withCachedTag(IndexLogEntryTags.INMEMORYFILEINDEX_INDEX_ONLY) {
+        new InMemoryFileIndex(spark, index.content.files, Map(), None)
+      }
+
+      new IndexHadoopFsRelation(
+        location,
+        new StructType(),
+        StructType(index.schema.filter(relation.schema.contains(_))),
+        if (useBucketSpec) Some(index.bucketSpec) else None,
+        new ParquetFileFormat,
+        Map(IndexConstants.INDEX_RELATION_IDENTIFIER))(spark, index)
+    }
+
     // Note that we transform *only* the base relation and not other portions of the plan
     // (e.g., filters). For instance, given the following input plan:
     //        Project(A,B) -> Filter(C = 10) -> Scan (A,B,C,D,E)
     // in the presence of a suitable index, the getIndexPlan() method will emit:
     //        Project(A,B) -> Filter(C = 10) -> Index Scan (A,B,C)
     plan transformDown {
       case baseRelation @ LogicalRelation(_: HadoopFsRelation, baseOutput, _, _) =>
-        val location = index.withCachedTag(IndexLogEntryTags.INMEMORYFILEINDEX_INDEX_ONLY) {
-          new InMemoryFileIndex(spark, index.content.files, Map(), None)
-        }
-
-        val relation = new IndexHadoopFsRelation(
-          location,
-          new StructType(),
-          StructType(index.schema.filter(baseRelation.schema.contains(_))),
-          if (useBucketSpec) Some(index.bucketSpec) else None,
-          new ParquetFileFormat,
-          Map(IndexConstants.INDEX_RELATION_IDENTIFIER))(spark, index)
-
+        val relation = makeHadoopFsRelation(index, baseRelation)
         val updatedOutput =
           baseOutput.filter(attr => relation.schema.fieldNames.contains(attr.name))
         baseRelation.copy(relation = relation, output = updatedOutput)
+      case v2Relation @ DataSourceV2Relation(_, output, _, _, _) =>
+        val relation = makeHadoopFsRelation(index, v2Relation)
+        val updatedOutput =
+          output.filter(attr => relation.schema.fieldNames.contains(attr.name))
+        new LogicalRelation(relation, updatedOutput, None, false)
```

Comment:
Do we lose anything by going from `DataSourceV2Relation` to `LogicalRelation`?

Reply:
The intention is to overwrite the plan to the relation that the index dataset uses. The index uses Parquet files which are not related to DataSourceV2 API in any way.